index.coffee

```diff
@@ -1,60 +1,161 @@
-through = require 'through'
+stream = require 'readable-stream'
+throttle = require 'lodash.throttle'
 
 module.exports = (requestModule, requestOpts, opts) ->
-	res = new Resumable(requestModule, requestOpts, opts)
-	return res.resume()
-
-class Resumable
-	constructor: (requestModule, requestOpts, opts) ->
-		@stream = through (data) =>
-			@bytesRead += data.length
-			@stream.emit('data', data)
+	return new ResumableRequest(requestModule, requestOpts, opts)
+
+module.exports.defaults = (defaults = {}) ->
+	(requestModule, requestOpts, opts) ->
+		opts = Object.assign {}, defaults, opts
+		return new ResumableRequest(requestModule, requestOpts, opts)
+
+cloneResponse = (response) ->
+	# creates an `http.IncomingMessage`-like stream, that can also be piped to.
+	s = new stream.PassThrough()
+	[ 'headers', 'httpVersion', 'method', 'rawHeaders', 'statusCode', 'statusMessage' ].forEach (prop) ->
+		s[prop] = response[prop]
+	return s
+
+wrapLegacyStream = (s) ->
+	new stream.Readable().wrap(s)
+
+class ResumableRequest extends stream.Readable
+	constructor: (@requestModule, @requestOpts, opts = {}) ->
+		super()
+
+		@error = null
+		@request = null
+		@response = null
+
 		@bytesRead = 0
+		@bytesTotal = null
 		@retries = 0
-		@maxRetries = opts?.maxRetries ? 10
-		@contentLength = null
-		@requestOpts = requestOpts
-		@requestModule = requestModule
-		@error = null
+		@maxRetries = opts.maxRetries ? 10
+
+		@_retry = throttle(@_retry.bind(this), opts.retryInterval ? 1000)
+		@_reportProgress = throttle(@_reportProgress.bind(this), opts.progressInterval ? 1000)
+
+		@on 'pipe', =>
+			# request can be written to when uploading a resource to a URL.
+			# We do not support that -- resumable request can only be used
+			# to *download* a resource, i.e. only the response is resumable.
+			# piping will fail with a not-so-obvious error as soon as the stream
+			# gets going because we don't implement Writable, so it's detect this
+			# and emit a more appropriate error.
+			@error = new Error('ResumableRequest is not writable')
+			# have to abort asynchronously to allow client code to register 'error' handlers
+			process.nextTick =>
+				@_retry()
+
+		@_request()
+
+	abort: ->
+		@emit('abort')
+		@_end()
+
+	push: (data, encoding) ->
+		@retries = 0 # we got some data, reset number of retries
+		@bytesRead += data.length if data?
+		@_reportProgress()
+		return super(data, encoding)
+
+	_end: ->
+		return if @ended
+		@ended = true
+		doEnd = =>
+			@request.abort()
+			@_retry.cancel()
+			@_reportProgress()
+			@_reportProgress.flush()
+			@_reportProgress.cancel()
+			@emit('end')
+			if not @error?
+				@emit('complete')
+		if @response?
+			@response.end(doEnd)
+		else
+			doEnd()
+
+	_read: -> # noop -- we're manually pushing buffers into self
```

Comment:
@dfunckt oh I see, you're right – just read through the `request` source, and it's not even an old-style stream, but inherits from `stream.Stream` (ideally it would be a `stream.Duplex`) and doesn't implement the `_read()` at all – it just spews out what it gets with `this.emit('data', chunk)`... which is probably not what you want (as that doesn't care about back-pressure either).

Might be good to use [`simple-get`](https://github.com/feross/simple-get) instead, which is a proper stream, and works like a charm – runnable example with the code from above: https://gist.github.com/jhermsmeier/1e21fdd4ede36660226f0b163d4ef1d6

Regarding the PassThrough handling the backpressure; I don't think that works, as the source (in this case the actual `http.IncomingMessage`) needs to get the signal (meaning only be read from when needed) – and since `request`'s internals clearly don't handle this properly, it'll just all end up in the PassThrough stream's input buffer and balloon memory usage.